### PR TITLE
Copy endpoint info, runtime info, commmand parsing, and tests from dotnet/diagnostics

### PR DESCRIPTION
**Pull request by jander-msft**
These changes move some source from the diagnostics repo to dotnet-monitor. This is done because the endpoint info (and related) classes are not used by the other diagnostic tools and allows dotnet-monitor to eventually refactor some of the abstracts into simpler forms. This change is the first step of just porting the code from diagnostics to dotnet-monitor without changing any of the functionality. Some of the test code was changed in order to fit the existing test expectations in this repository.

cc @josalem, @hoyosjs

### CODE REVIEW

**Comment by hoyosjs** [comment]
I am happy with this living here. However, does this affect in any way the triggers + channels functionality that was proposed? The ASP.NET Core library that was proposed at some point could live here with no issues.

**Comment by jander-msft** [comment]
> I am happy with this living here. However, does this affect in any way the triggers + channels functionality that was proposed? The ASP.NET Core library that was proposed at some point could live here with no issues.

Chatted offline, but putting it here for everyone else. The pipeline implementations will remain in the diagnostics repo. The triggers feature will be partially implemented in the diagnostics repo (the pipeline and trigger conditions) and partially in dotnet-monitor (configuration of trigger conditions with actions and filters). The web API library will likely remain in the dotnet-monitor repo.

File: src/Tests/Microsoft.Diagnostics.Monitoring.UnitTests/EndpointInfoSourceTests.cs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+            appRunner.ConnectionMode = Options.DiagnosticPortConnectionMode.Connect;
+            appRunner.DiagnosticPortPath = transportName;
+            appRunner.ScenarioName = TestAppScenarios.AsyncWait.Name;
+            return appRunner;
+        }
+
+        private async Task<IEnumerable<IEndpointInfo>> GetEndpointInfoAsync(ServerEndpointInfoSource source)
+        {
+            _outputHelper.WriteLine("Getting endpoint infos.");
+            using CancellationTokenSource cancellationSource = new(TimeSpan.FromSeconds(10));
```
Comment (wiktork) [inline src/Tests/Microsoft.Diagnostics.Monitoring.UnitTests/EndpointInfoSourceTests.cs:219]: Make into constant?

File: src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/Microsoft.Diagnostics.Monitoring.TestCommon.csproj [hunk 2]
```diff
@@ -2,11 +2,17 @@
...
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Include="..\..\Tools\dotnet-monitor\DiagnosticPortConnectionMode.cs" Link="Options\DiagnosticPortConnectionMode.cs" />
+  </ItemGroup>
+
   <ItemGroup>
     <PackageReference Include="System.Security.Principal.Windows" Version="$(SystemSecurityPrincipalWindowsVersion)" />
     <PackageReference Include="Microsoft.DotNet.XUnitExtensions" Version="$(MicrosoftDotNetXUnitExtensionsVersion)" />
+    <PackageReference Include="xunit.assert" Version="$(XunitAssertVersion)" />
```
Comment (wiktork) [inline src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/Microsoft.Diagnostics.Monitoring.TestCommon.csproj:15]: I remember having this issue, and certain frameworks were complaining that xunitassert is already part of the framework. How did you get around that issue?

**Review by wiktork — APPROVED** [review]

File: src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/Microsoft.Diagnostics.Monitoring.TestCommon.csproj (hunk 2 above)
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/Microsoft.Diagnostics.Monitoring.TestCommon.csproj:15]: This one isn't named "UnitTests". Anecdotally, I've only seen it automatically treat a project as a unit test project if it has "UnitTests" in its name.

**Pull request merged**